config/config-mp/src/main/java/io/helidon/config/mp/MpConfigProviderResolver.java

```diff
@@ -173,13 +173,23 @@ public void releaseConfig(Config config) {
         Lock lock = RW_LOCK.readLock();
         AtomicReference<ClassLoader> cl = new AtomicReference<>();
 
+        // in case we get our own delegate, we want to remove the exact same instance of delegate
         try {
             lock.lock();
-            for (Map.Entry<ClassLoader, ConfigDelegate> entry : CONFIGS.entrySet()) {
-                Config configFromRef = entry.getValue().delegate();
-                if (config == configFromRef) {
-                    cl.set(entry.getKey());
-                    break;
+            if (config instanceof ConfigDelegate) {
+                for (Map.Entry<ClassLoader, ConfigDelegate> entry : CONFIGS.entrySet()) {
+                    if (config == entry.getValue()) {
+                        cl.set(entry.getKey());
+                        break;
+                    }
+                }
+            } else {
+                for (Map.Entry<ClassLoader, ConfigDelegate> entry : CONFIGS.entrySet()) {
+                    Config configFromRef = entry.getValue().delegate();
+                    if (config == configFromRef) {
+                        cl.set(entry.getKey());
+                        break;
+                    }
```

Comment:
Maybe something like this could avoid code duplication
```java
Predicate<Map.Entry<ClassLoader, ConfigDelegate>> entryPredicate;

if(config instanceof ConfigDelegate){
    entryPredicate = entry -> config == entry.getValue();
}else {
    entryPredicate = entry -> config == entry.getValue().delegate();
}

CONFIGS.entrySet()
        .stream()
        .filter(entryPredicate)
        .findFirst()
        .map(Map.Entry::getKey)
        .ifPresent(cl::set);
```

Reply:
On the other hand it's not shorter nor more readable and definitely not faster